source/common/router/rds_impl.cc

```diff
@@ -154,21 +155,24 @@ RouteConfigProviderManagerImpl::routeConfigProviders() {
 };
 
 Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::getRouteConfigProvider(
-    const Json::Object& config, Upstream::ClusterManager& cm, Stats::Scope& scope,
+    const envoy::api::v2::filter::Rds& rds, Upstream::ClusterManager& cm, Stats::Scope& scope,
     const std::string& stat_prefix, Init::Manager& init_manager) {
 
-  // RdsRouteConfigProviders are unique based on their <route_config_name>MAP_CONCATENATOR<cluster>.
+  // RdsRouteConfigProviders are unique based on their
+  // <route_config_name>MAP_CONCATENATOR<serialized RDS config>.
+  // TODO(htuch): Full serialization here gives large IDs, could get away with a
+  // strong hash instead.
   const std::string manager_identifier =
-      config.getString("route_config_name") + MAP_CONCATENATOR + config.getString("cluster");
+      rds.route_config_name() + MAP_CONCATENATOR + rds.SerializeAsString();
```

Comment:
Yeah, that was me just mechanically replacing things, thanks for pointing this out. I'm wondering about whether we have concerns if we use `std::hash` here - it's not crypto strong, the standard doesn't impose any strength requirements on the hash. So far, if we get collisions in xDS update hashes, we can miss updates, in this case we can actually get the wrong route table. Should we use a strong hash from BoringSSL or something? It seems using weak hashes is only legit if you also do a compare against the original strings to ensure there are no collisions, as you would in a simple hash table implementation for example. But, we don't keep these around if we're hashing as we do today.